lib/clAdapters/googleAdapter.js

```diff
@@ -1,37 +1,386 @@
 'use strict';
 
-var util = require('util'),
-  BaseAdapter = require('./baseAdapter');
-
-/**
- * GoogleAdapter
- *
- * `credentials` format:
- * ```
- * {
- *   email: 'test@test.com',
- *   password: 'password',
- *   account: 123456,
- *   role: 3
- * }
- * ```
- * @class
- * @return {GoogleAdapter}
- */
+var uuid = require('node-uuid');
+var crypto = require('crypto');
+var rp = require('request-promise');
+var util = require('util');
+var BaseAdapter = require('./baseAdapter');
+var GoogleMail = require('./google-js.js');
+var moment = require('moment');
+var querystring = require('querystring');
+var _ = require('lodash');
+require('es6-promise').polyfill();
+
 var GoogleAdapter = module.exports = function GoogleAdapter() {
   BaseAdapter.call(this);
 };
 
 util.inherits(GoogleAdapter, BaseAdapter);
 
-/**
- * @override
- */
 GoogleAdapter.prototype.init = function() {
+  var _this = this;
+  this._config = new GoogleMail.Configuration(this.credentials);
+  this._service = new GoogleMail.Service(this._config);
+  return this._service
+    .init()
+    .then(function( /*client*/ ) {
+      var msg = 'Successfully initialized gmail adapter for email: %s';
+      console.log(msg, _this.credentials.email);
+      return _this;
+    });
 };
 
-/**
- * @override
- */
 GoogleAdapter.prototype.reset = function() {
+  delete this._config;
+  delete this._service;
+};
+
+var getSingleMessageDetails = function(messageId, userEmail, token, apiVersion, additionalFields, result) {
+  var additionalFieldsToQuery = additionalFields.replace('BodyPreview', 'snippet');
+  additionalFieldsToQuery = additionalFieldsToQuery.replace('Body', 'payload(parts)');
+  additionalFieldsToQuery = additionalFieldsToQuery.replace('Subject', 'payload(parts)');
+
+  if(additionalFieldsToQuery != '') {
+    additionalFieldsToQuery = ',' + additionalFieldsToQuery;
+  }
+
+  var messageRequestOptions = {
+    method: 'GET',
+    uri: 'https://www.googleapis.com/gmail/v' + apiVersion + '/users/' + userEmail + '/messages/' + messageId + '?fields=id,threadId,labelIds,payload(headers)' + additionalFieldsToQuery,
+    headers : {
+      Authorization: 'Bearer ' + token,
+      Accept: 'application/json;odata.metadata=none'
+    }
+  };
+  return rp(messageRequestOptions)
+  .then(function(messageDetails) {
+    result.messageData = JSON.parse(messageDetails);
+    if(additionalFields.indexOf('Subject') == -1) {
+      //remove subject header
+      if(result.messageData.payload && result.messageData.payload.headers && result.messageData.payload.headers.length > 0) {
+        for(var headerIter = 0; headerIter < result.messageData.payload.headers.length; headerIter++) {
+          if(result.messageData.payload.headers[headerIter].name === 'Subject') {
```

Comment:
Future: potentially make this a config option so application can decide whether to pull in Subject (not blocking)